stdlib/public/core/NormalizedCodeUnitIterator.swift

```diff
@@ -0,0 +1,253 @@
+  }
+  
+  init(_ guts: _StringGuts, startIndex: Int = 0) {
+    source = StringGutsSource(guts, start: startIndex)
+  }
+  
+  mutating func compare(with other: NormalizedCodeUnitIterator) -> _Ordering {
+    var mutableOther = other
+    for cu in IteratorSequence(self) {
+      if let otherCU = mutableOther.next() {
+        let result = _lexicographicalCompare(cu, otherCU)
+        if result == .equal {
+          continue
+        } else {
+          return result
+        }
+      } else {
+        //other returned nil, we are greater
+        return .greater
+      }
+    }
+    
+    //we ran out of code units, either we are equal, or only we ran out and
+    //other is greater
+    if let _ = mutableOther.next() {
+      return .less
+    } else {
+      return .equal
+    }
+  }
+  
+  struct UnmanagedOpaqueStringSource: SegmentSource {
+    var remaining: Int {
+      return opaqueString.count - index
+    }
+    var opaqueString: _UnmanagedOpaqueString
+    var index: Int
+    
+    init(_ opaqueString: _UnmanagedOpaqueString, start: Int = 0) {
+      self.opaqueString = opaqueString
+      index = start
+    }
+    
+    mutating func tryFill(buffer: UnsafeMutableBufferPointer<UInt16>) -> Int? {
+      var bufferIndex = 0
+      let originalIndex = index
+      repeat {
+        guard index < opaqueString.count else {
+          break
+        }
+        
+        guard bufferIndex < buffer.count else {
+          //The buffer isn't big enough for the current segment
+          index = originalIndex
+          return nil
+        }
+        
+        let cu = opaqueString[index]
+        buffer[bufferIndex] = cu
+        index += 1
+        bufferIndex += 1
+      } while !opaqueString.hasNormalizationBoundary(after: index - 1)
+      
+      return bufferIndex
+    }
+  }
+  
+  struct UnmanagedStringSource: SegmentSource {
+    var remaining: Int {
+      return unmanagedString.count - index
+    }
+    
+    var unmanagedString: _UnmanagedString<UInt16>
+    var index: Int
+    
+    init(_ unmanagedString: _UnmanagedString<UInt16>, start: Int = 0) {
+      self.unmanagedString = unmanagedString
+      index = start
+    }
+    
+    mutating func tryFill(buffer: UnsafeMutableBufferPointer<UInt16>) -> Int? {
+      var bufferIndex = 0
+      let originalIndex = index
+      repeat {
+        guard index < unmanagedString.count else {
+          break
+        }
+        
+        guard bufferIndex < buffer.count else {
+          //The buffer isn't big enough for the current segment
+          index = originalIndex
+          return nil
+        }
+        
+        let cu = unmanagedString[index]
+        buffer[bufferIndex] = cu
+        index += 1
+        bufferIndex += 1
+      } while unmanagedString.hasNormalizationBoundary(
+          after: index - 1, 
+          count: unmanagedString.count) == false
+      
+      return bufferIndex
+    }
+  }
+  
+  struct StringGutsSource: SegmentSource {
```

Comment:
Probably underscore the name.

I think we'll adjust this approach later this release, either in favor of just unmanaged opaque string or a more generally useful streaming opaque string, but that's pending some other work.